### ISSUE TITLE
Unable to get the logs of services on docker swarm (docker -v 17.06.2-ce)

### ISSUE THREAD

**Issue by nicolaselkhoury** [moby/moby#35932]
**Description**
We have a docker swarm composed of three master nodes (drained), and four minion nodes. We have a set of services running on the swarm, where each service has three replicas. The docker version of each node is 17.06.2-ce.

**Steps to reproduce the issue:**
1. Create a Docker swarm of three drained master nodes and four minion nodes.
2. Create a bunch of services, where each service is composed of three replicas
3. docker service logs <service name>

**Describe the results you received:**
No output

**Describe the results you expected:**
I am expecting to list the outputs of all the containers of a particular service

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**
```
17.06.2-ce
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 17.06.2-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 36
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: tae3ntjkjafz6o8gilrivqcx4
 Is Manager: true
 ClusterID: xh6n8d2wss8xtk53d7camhoy8
 Managers: 3
 Nodes: 7
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: <HOST3>
 Manager Addresses:
 <HOST1>:2377
  <HOST2>:2377
  <HOST3>:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-1038-aws
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.858GiB
Name:<NAME>
ID: SNVJ:LV7E:FWTF:5KGV:INZ3:SFLX:FI5K:UBVL:XAQP:6RIG:KOSV:JSE3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS EC2 ubuntu:16.04 machines

**Comment by thaJeztah** [comment]
Please provide more information, as there's not a lot to go on;

- are logs _never_ working, or after some time?
- are you using the default (`json-file`) logging driver for those services? what logging options show if you `docker inspect` a container for the service?
- what's running in those services? are the services logging to `stdout`/`stderr` (i.e., if you `docker container logs <container-id>` on the node where the container is running; are logs shown?
- is there anything in the daemon logs (check both the manager's logs, and the worker node) that provides something useful?

Be aware that Docker 17.06 has reached end-of-life in October, and is no longer maintained, so will no longer receive updates. Consider updating to a current version of Docker or if you want to stay on the 17.06 version, switch to Docker Enterprise Edition, which still maintains 17.06

**Comment by nicolaselkhoury** [comment]
Thank you for your prompt response, I will answer your questions in the same order.

- We have the cluster running for months, and the logs disappeared a few days ago. Before that, we were consistently able to get the service logs.

- The logging options for all containers of each service are using the default json-file logging driver.

- Yes, a docker logs <container-id> generates the corrects logs of a container.

- No I could not find anything interesting in the daemon logs, unless there is something I am missing.

Is there a stable release with long term support that we could use? How easy is it to upgrade a swarm to a newer version without any downtime of the swarm services?

**Comment by thaJeztah** [comment]
Thanks for the extra information. I know there's still a number of tickets tracking issues around this;

- https://github.com/moby/moby/issues/33203 ("Service Logs fails on scale down")
- https://github.com/moby/moby/issues/33183 ("`docker service logs` stops showing logs from containers on different nodes")

I don't recall if a cause was found for the above (and could be the issue you're seeing here).

One improvement that is still on the roadmap (not a solution for the bug, but related) is to allow logs to be sent to multiple logging drivers (see https://github.com/moby/moby/issues/17910).

The intent of the `docker service logs` option was to have an easy/quick way to check service logs in the swarm, but it was never really intended as a full log-management solution; in a cluster setup, nodes are usually "ephemeral", so it's important to get logs offloaded from the nodes where possible (and aggregated centrally). Currently, aggregating logs usually requires a logging-driver, other than `json-file`, but in that case, logs can no longer be viewed using `docker service logs`; being able to send logs to multiple drivers (or have a temporary buffer for the last X log-entries for viewing) would resolve that.

> Is there a stable release with long term support that we could use?

The "stable" channel for Docker Community Edition is maintained for 4 months; Docker 17.03-CE, 17.09, and 17.12 are "stable" releases, but 3 months is not LTS. Docker EE (enterprise edition) does offer 12 (or 18 months? I should check) of support for releases.

> How easy is it to upgrade a swarm to a newer version without any downtime of the swarm services?

This depends a bit on your situation (e.g., are you running stateful services; if so, where / how is data stored? Can services be re-scheduled to another node without problems? Do all services have more than one replica?).

When updating, it's recommended to not "skip" versions; so update from 17.06 to 17.09 to 17.12 if possible. Update one node at a time, and make sure that other nodes have enough resources to re-schedule existing services. Another approach is to _add_ a new node (with an updated version of docker), then de-commision an old node (you can drain the old node after the new node was joined).

Take care with updating managers to prevent loosing quorum during the update (never, ever, try updating them all at the same time :smile:, and make sure managers successfully re-join the cluster after updating).

TL;DR; it depends on your situation/setup, and be sure to try on a non-production cluster first, and have a backup plan (which is just common practice :smile:)

**Comment by thaJeztah** [comment]
Actually; meant to link to this issue, which is tracking the roadmap/enhancements https://github.com/moby/moby/issues/33475 ("[Proposal] Enhance local logging / remote logging")

**Comment by nicolaselkhoury** [comment]
Thank you very much for your replies. They are extremely helpful. I only have one concern, and I will rephrase the question so that it can be understood better. 

We have a cluster of 3 manager nodes, and 4 minion nodes. 
Assume we have 5 services of 3 replicas each. 
If I were to upgrade the nodes, one by one, version by version, as you suggested, will an updated node be able to join the cluster of nodes (which clearly have the older 17.06 version, without problems? Of course I will test this on another cluster before messing with this one haha.

Will the update potentially fix the logging issue? or it is not necessary?

**Comment by thaJeztah** [comment]
Yes, the new nodes (new version of Docker) should be able to re-join the existing cluster. 

> Will the update potentially fix the logging issue?

Unfortunately, I'm not sure if it will resolve this particular issue. Perhaps @dperny knows of changes in this area, but I don't think the issues I linked to have been resolved (also because I think there was not a consistent way to reproduce the problem, which makes debugging really difficult).

**Comment by nicolaselkhoury** [comment]
Here's an interesting thing that we just found out. The problem is bigger than just gathering service logs. The swarm is not acting as a swarm. I am trying to update/create services. The swarm assigns containers to nodes, but nothing happens past that.

For instance, assume I am deploying a service of 3 replicas. All three replicas are assigned to their designated nodes. However, the current state remains "assigned"

**Comment by thaJeztah** [comment]
Looks like those logs are truncated; can you try getting the non-truncated logs?

**Comment by nicolaselkhoury** [comment]
@thaJeztah I updated my previous comment with a solution. Turns out, the leader somehow was messed up, without signaling any kind of error. Just by electing a new leader, the swarm is back to normal.

Thank you so much for your time and effort, and your prompt and quick replies.

**Comment by thaJeztah** [comment]
Thanks for the update; not sure what happened then (but good to hear the situation was resolved - at least for now).

Since you no longer have the situation present, I'll close this issue, but feel free to comment at a later stage if you have more information that could be useful for debugging 👍

**Comment by MGD1981** [comment]
This just happened to us -- demoting and re-promoting fixed it, so thanks for that.

**Comment by rbdiang** [comment]
This fixed our issue too
